keps/sig-node/4817-resource-claim-device-status/README.md

```diff
@@ -0,0 +1,730 @@
+      variables.userNodeName != variables.objectNodeName
+    messageExpression: >-
+      "User '" + request.userInfo.username + "' on node '" + variables.userNodeName + "' is not allowed to update the .status.devices of a ResourceClaim allocated on node '" + variables.objectNodeName + "'."
+    reason: Forbidden
+  variables:
+  - name: userNodeName
+    expression: >-
+      request.userInfo.extra[?'authentication.kubernetes.io/node-name'][0].orValue('')
+  - name: objectNodeName
+    expression: >-
+      object.status.allocation.nodeSelector.nodeSelectorTerms[0].matchFields[0].values[0].orValue('')
+---
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicyBinding
+metadata:
+  name: "resourceclaim-device-status-update-binding"
+spec:
+  policyName: "resourceclaim-device-status-update"
+  validationActions: [Deny]
+```
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- `ResourceClaim` validation:
+    - A device can only be reported once in the `ResourceClaim`.
+    - The reported device is allocated in the `ResourceClaim`.
+    - Properties set in `AllocatedDeviceStatus` are in the correct format.
+- `ResourceClaim` registry strategy:
+    - With the feature gate enabled, the field is not dropped.
+    - With the feature gate disabled, the field is dropped.
+
+Coverage:
+- `k8s.io/kubernetes/pkg/apis/resource/validation`: `9/28/2024` - `77.1`
+- `k8s.io/kubernetes/pkg/registry/resource/resourceclaim`: `9/28/2024` - `75.5`
+
+##### Integration tests
+
+- Usage of the `Devices` field in the `ResourceClaimStatus`:
+    * With the feature gate enabled, the field exists in the `ResourceClaim`.
+    * With the feature gate disabled, the field does not exist in the
+      `ResourceClaim`.
+    * With the feature gate enabled, the `ValidatingAdmissionPolicy` exists and
+      restricts the write access of the `ResourceClaim.Status.Devices`.
+
+##### e2e tests
+
+The [DRA test driver](https://github.com/kubernetes/kubernetes/tree/master/test/e2e/dra/test-driver)
+will be extended to support the new  `ResourceClaim.Status.Devices` field and to 
+populate it with data related to the device configured in the pod.
+
+A new network DRA Driver will be implemented (or extended from the existing DRA 
+test driver) to support networking type of devices and report their 
+network status.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind feature gates (`DRAResourceClaimDeviceStatus`).
+  Feature Gates are disabled by default.
+- Documentation provided.
+- Initial unit, integration and e2e tests completed and enabled.
+- Authorization implemented to allow only the user on the same node as the 
+  allocated `ResourceClaim` to write the status of the devices.
+
+#### Beta
```

Comment:
Talked with @danwinship offline, @LionelJouin @dougbtv what do you think if we add to beta criteria having a multus DRA network driver ?

Reply:
I'd definitely be interested in it! And there's a sort of basis that exists pre-1.31 in: https://github.com/k8snetworkplumbingwg/multus-cni/pull/1078

But yes, it'd need extensions to this functionality as well as a Multus DRA driver itself, as the challenge with that implementation is that it uses an example reference driver, and directly being a driver would be a good way to see how well the spec fits with an implementation.

Reply:
I added 1 real-world usage as beta requirement.